src/languages/clickhouse/clickhouse.formatter.ts

```diff
@@ -0,0 +1,161 @@
+  'EXECUTE AS',
+  // https://clickhouse.com/docs/sql-reference/statements/use
+  'USE',
+  // https://clickhouse.com/docs/sql-reference/statements/move
+  'MOVE {USER | ROLE | QUOTA | SETTINGS PROFILE | ROW POLICY}',
+  // https://clickhouse.com/docs/sql-reference/statements/check-grant
+  'CHECK GRANT',
+  // https://clickhouse.com/docs/sql-reference/statements/undrop
+  'UNDROP TABLE',
+]);
+const tabularOnelineClauses = expandPhrases([
+  // https://clickhouse.com/docs/sql-reference/statements/create
+  'CREATE {DATABASE | NAMED COLLECTION} [IF NOT EXISTS]',
+  'CREATE [OR REPLACE] {VIEW | DICTIONARY} [IF NOT EXISTS]',
+  'CREATE MATERIALIZED VIEW [IF NOT EXISTS]',
+  'CREATE FUNCTION',
+  'CREATE {USER | ROLE | QUOTA | SETTINGS PROFILE} [IF NOT EXISTS | OR REPLACE]',
+  'CREATE [ROW] POLICY [IF NOT EXISTS | OR REPLACE]',
+  // https://clickhouse.com/docs/sql-reference/statements/create/table#replace-table
+  'REPLACE [TEMPORARY] TABLE [IF NOT EXISTS]',
+  // https://clickhouse.com/docs/sql-reference/statements/alter
+  'ALTER [TEMPORARY] TABLE',
+  'ALTER {USER | ROLE | QUOTA | SETTINGS PROFILE} [IF EXISTS]',
+  'ALTER [ROW] POLICY [IF EXISTS]',
+  'ALTER NAMED COLLECTION [IF EXISTS]',
+  // https://clickhouse.com/docs/sql-reference/statements/delete
+  'DELETE FROM',
+  // https://clickhouse.com/docs/sql-reference/statements/grant
+  'GRANT [ON CLUSTER]',
+  // https://clickhouse.com/docs/sql-reference/statements/revoke
+  'REVOKE [ON CLUSTER]',
+  // https://clickhouse.com/docs/sql-reference/statements/check-table
+  'CHECK TABLE',
+  // https://clickhouse.com/docs/sql-reference/statements/describe-table
+  '{DESC | DESCRIBE} TABLE',
+]);
+
+const reservedSetOperations = expandPhrases([
+  // https://clickhouse.com/docs/sql-reference/statements/select/set-operations
+  'UNION [ALL | DISTINCT]',
+  // https://clickhouse.com/docs/sql-reference/statements/parallel_with
+  'PARALLEL WITH',
+]);
+
+const reservedJoins = expandPhrases([
+  // https://clickhouse.com/docs/sql-reference/statements/select/join
+  '[GLOBAL] [INNER|LEFT|RIGHT|FULL|CROSS] [OUTER|SEMI|ANTI|ANY|ALL|ASOF] JOIN',
+]);
+
+// https://clickhouse.com/docs/sql-reference/syntax
+export const clickhouse: DialectOptions = {
+  name: 'clickhouse',
+  tokenizerOptions: {
+    reservedSelect,
+    reservedClauses: [...reservedClauses, ...standardOnelineClauses, ...tabularOnelineClauses],
+    reservedSetOperations,
+    reservedJoins,
+    reservedKeywordPhrases: keywordPhrases,
+
+    reservedKeywords: keywords,
+    reservedDataTypes: dataTypes,
+    reservedFunctionNames: functions,
+    nestedBlockComments: false,
+    underscoresInNumbers: true,
+    stringTypes: ['$$', "''-qq", "''-qq-bs"],
+    identTypes: ['""-qq', '``'],
+    paramTypes: {
+      // https://clickhouse.com/docs/sql-reference/syntax#defining-and-using-query-parameters
+      custom: [
+        {
+          regex: String.raw`\{\s*[a-zA-Z0-9_]+\s*:\s*[a-zA-Z0-9_]+\s*\}`,
```

Comment:
That regex would fail when a more complex type is used in the parameter, like this:

```sql
SELECT {d: Map(String, Array(UInt8))};
```